src/functional/ffront/dialect_parser.py

```diff
@@ -0,0 +1,166 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import ast
+import textwrap
+import types
+from dataclasses import dataclass
+from typing import Any, ClassVar, Generic, Optional, Type, TypeVar, cast
+
+from eve.type_definitions import SourceLocation
+from functional import common
+from functional.ffront.source_utils import ClosureRefs, SourceDefinition, SymbolNames
+
+
+DialectRootT = TypeVar("DialectRootT")
+
+
+@dataclass(frozen=True, kw_only=True)
+class DialectParser(ast.NodeVisitor, Generic[DialectRootT]):
+    source: str
+    filename: str
+    starting_line: int
+    closure_refs: ClosureRefs
+    externals_defs: dict[str, Any]
+
+    syntax_error_cls: ClassVar[Type[DialectSyntaxError]]
+
+    @classmethod
+    def apply(
+        cls,
+        source_definition: SourceDefinition,
+        closure_refs: ClosureRefs,
+        externals: Optional[dict[str, Any]] = None,
+    ) -> DialectRootT:
+        source, filename, starting_line = source_definition
+        try:
+            definition_ast = cls._preprocess_definition_ast(
+                ast.parse(textwrap.dedent(source)).body[0]
+            )
+            output_ast = cls._postprocess_dialect_ast(
+                cls(
+                    source=source,
+                    filename=filename,
+                    starting_line=starting_line,
+                    closure_refs=closure_refs,
+                    externals_defs=externals or {},
+                ).visit(definition_ast)
+            )
+            cls._assert_source_invariants(source_definition, closure_refs)
```

Comment:
I would call it like:
```
if __debug__:
    assert_source_invariants(source_definition, closure_refs)
```
Check here for meaning of `__debug__`, https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement
See other comment about moving this function out of the class.